bitnami/etcd/templates/preupgrade-hook-job.yaml

```
@@ -0,0 +1,180 @@
+{{- /*
+Copyright Broadcom, Inc. All Rights Reserved.
+SPDX-License-Identifier: APACHE-2.0
+*/}}
+apiVersion: {{ include "common.capabilities.job.apiVersion" . }}
+kind: Job
+metadata:
+  name: {{ include "common.names.fullname" . }}-pre-upgrade
+  namespace: {{ .Release.Namespace | quote }}
+  labels: {{- include "common.labels.standard" ( dict "customLabels" .Values.commonLabels "context" $ ) | nindent 4 }}
+    app.kubernetes.io/component: etcd-pre-upgrade-job
+  annotations: {{- include "common.tplvalues.render" (dict
+    "value" (merge 
+      .Values.commonAnnotations
+      (dict
+        "helm.sh/hook" "pre-upgrade"
+        "helm.sh/hook-delete-policy" "before-hook-creation")
+    )
+    "context" $) | nindent 4 }}
```

Please use the alternative below:
```suggestion
  {{- $defaultAnnotations := dict "helm.sh/hook" "pre-upgrade" "helm.sh/hook-delete-policy" "hook-succeeded" }}
  {{- $annotations := include "common.tplvalues.merge" ( dict "values" ( list .Values.commonAnnotations $defaultAnnotations ) "context" . ) }}
  annotations: {{- include "common.tplvalues.render" ( dict "value" $annotations "context" $ ) | nindent 4 }}
```
Please note I switched the delete policy given I don't think the job is required after it succeeds.

---

[Argo CD docs](https://argo-cd.readthedocs.io/en/latest/user-guide/helm/#hook-tips) recommend setting this value to `before-hook-creation` to make sure it runs on every sync. I think it's a sensible default. Perhaps, we should allow customization via Helm values?

---

Feel free to keep it with `before-hook-creation`. Actually, it's good to keep the "completed" pods to inspect the logs under certain circumstances.
Please apply the other suggestion for using `common.tplvalues.merge` to merge annotations.

---

Suggestion applied ;)